### DISCUSSION TITLE
EntitySpawnEvent with optional source ItemStack

### DISCUSSION THREAD

**Discussion by Prof-Bloodstone** [PaperMC/Paper#10498] [Ideas]
**Is your feature request related to a problem? Please describe.**
I'd love to be able to modify the entity when it gets spawned from some ItemStack (spawn egg, boat, item frame, ...) and act on that.

**Describe the solution you'd like**
Ability to retrieve to retrieve source ItemStack which caused the Entity to spawn.
I think it'd make most sense to add it as nulllable value to existing EntitySpawnEvent, since other events like CreatureSpawnEvent and ProjectileLaunchEvent might potentially also use this.

**Describe alternatives you've considered**
Patching the server source code directly :/

**Comment by Proximyst** [comment]
I've looked into this before, and it's a tedious process to say the least. PRs welcome, but not a beginner issue :p

**Comment by Machine-Maker** [comment]
There's no overriden method for example to create an entity from an item, so this would probably have to be done for each item that does this which requires a list of items that do it
* Spawn Eggs
* Boats
* ~~Item Frame~~ (HangingPlaceEvent)
* ~~Painting~~ (HangingPlaceEvent)
* ~~Leash~~ (HangingPlaceEvent)
* Armor Stand
* End Crystals
* Fish Buckets
* Fireworks

and what am I missing?
